lib/pbench/server/api/resources/query_apis/datasets_publish.py

```diff
@@ -0,0 +1,207 @@
+
+        {
+            "name": "dataset name",
+            "controller": "controller name",
+            "user": "username",
+            "access": "desired access"
+        }
+
+        json_data: JSON dictionary of type-normalized parameters
+            user: specifies the owner of the data to be modified; it need not
+                necessarily be the user represented by the session token
+                header, assuming the authenticated user can access "user"s
+                data.
+            controller: the controller that generated the dataset
+            name: name of the dataset to publish
+            access: The desired access level of the dataset (currently either
+                "private" or "public")
+
+        context: The Dataset record, which contains the root run-data index
+            document ID.
+
+        This payload makes use of the Elasticsearch bulk query syntax, which
+        allows combining multiple queries into a single JSON payload using the
+        "Content-Type: application/x-ndjson" header:
+
+        {"index": "jam-pbench.v6.run-data.2001-05"}\n
+        {"authorization.access": "public"}\n
+
+        We rely on the document ID map built by the indexer to find the many
+        Elasticsearch documents created by Pbench indexing. This avoids the
+        need to use multiple Elasticsearch queries here to determine the set
+        of documents on which to act.
+
+        TODO: If some of the bulk updates fail, we need a mechanism either to
+        RETRY or to ROLL BACK (which would in turn require a RETRY); TBD.
+        """
+        user = json_data["user"]
+        name = json_data["name"]
+        access = json_data["access"]
+
+        self.logger.info(
+            "Update access for dataset {} for user {} to {}, prefix {}",
+            name,
+            user,
+            access,
+            self.prefix,
+        )
+
+        map = json.loads(Metadata.get(dataset=context, key=Metadata.INDEX_MAP).value)
+
+        # Construct an "NDJSON" document for a bulk update of all Elasticsearch
+        # documents associated with the Dataset.
+        #
+        # NDJSON is a series of JSON documents separated by "newlines" (defined
+        # by Elasticsearch as cr-lf pairs). We'll construct stringified JSON in
+        # an array which is then gathered up using str.join().
+        #
+        # TODO: it'd probably be better to use the Elasticsearch bulk helper and
+        # pyesbulk here; but I don't want to duplicate the setup in indexer.py,
+        # it's not well generalized, and I don't want to add that refactoring to
+        # this PR.
+
+        cmds = []
+        for index, ids in map.items():
+            for id in ids:
+                x = (
+                    f'{{"update": {{"_index": "{index}", "_id": "{id}"}}}}\n'
+                    f'{{"doc": {{"authorization": {{"access": "{access}"}}}}}}'
+                )
+                cmds.append(x)
```

Comment:
Hmm; these all seem substantially less readable, especially with `black` idiosyncrasies thrown in, but I'll think about it.

Reply:
I concur.  The problem is that I find the doubled-`{`'s particularly unpleasant, and I was casting about for alternatives.

The one other one I came up with is kinda funny:
```
                x = (
                    f'["update": ["_index": "{index}", "_id": "{id}"]\n'
                    f'["doc": ["authorization": ["access": "{access}"]]]'
                )
                cmds.append(x.replace('[', '{').replace(']', '}'))
```
How's that grab ya?  😀  (Of course, you could use something else instead of the `[`'s and `]`'s...like some nice Unicode character, like `«` and `»`. 😉)

Reply:
Oh, I can't resist one more try...let's go old school:
```
                cmds.append(
                    (
                        '{"update": {"_index": "%s", "_id": "%s"}}\n'
        	        '{"doc": {"authorization": {"access": "%s"}}}'
                    )
                    % (index, id, access)
	        )
```
To us old C hands, that's pretty easy to read....  😉

Reply:
Wow, you're not letting this go; I guess the doubled braces bug you a lot more than they bug me. OK, um, I really prefer the f-string notation, but I may actually go with this one.

Reply:
I think I'll let go if you turn down this last suggestion.  🙂

I just couldn't quit while there were still alternatives available.  😁